packages/oft-evm/contracts/NativeOFTAdapter.sol

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * @title NativeOFTAdapter
+ * @dev NativeOFTAdapter is a contract that adapts native currency to the OFT functionality.
+ *
+ * @dev WARNING: ONLY 1 of these should exist for a given global mesh,
+ * unless you make a NON-default implementation of OFT and needs to be done very carefully.
+ * @dev WARNING: The default NativeOFTAdapter implementation assumes LOSSLESS transfers, ie. 1 native in, 1 native out.
+ */
+abstract contract NativeOFTAdapter is OFTCore {
+
+    error InsufficientMessageValue(uint256 provided, uint256 required);
+    error CreditFailed();
+
+    /**
+     * @param _localDecimals The decimals of the native on the local chain (this chain). 18 on ETH.
+     * @param _lzEndpoint The LayerZero endpoint address.
+     * @param _delegate The delegate capable of making OApp configurations inside of the endpoint.
+     */
+    constructor(
+        uint8 _localDecimals,
+        address _lzEndpoint,
+        address _delegate
+    ) OFTCore(_localDecimals, _lzEndpoint, _delegate) {}
+
+    /**
+     * @dev Returns the address of the native token
+     * @return The address of the native token.
+     */
+    function token() public pure returns (address) {
+        return address(0);
+    }
+
+    /**
+     * @notice Indicates whether the OFT contract requires approval of the 'token()' to send.
+     * @return bool indicating whether approval of underlying token implementation is required.
+     *
+     * @dev In the case of default NativeOFTAdapter, approval is not required.
+     */
+    function approvalRequired() external pure virtual returns (bool) {
+        return false;
+    }
+
+    /**
+     * @dev Executes the send operation while ensuring the correct amount of native is sent.
+     * @param _sendParam The parameters for the send operation.
+     * @param _fee The calculated fee for the send() operation.
+     *      - nativeFee: The native fee.
+     *      - lzTokenFee: The lzToken fee.
+     * @param _refundAddress The address to receive any excess funds.
+     * @return msgReceipt The receipt for the send operation.
+     * @return oftReceipt The OFT receipt information.
+     *
+     * @dev MessagingReceipt: LayerZero msg receipt
+     *  - guid: The unique identifier for the sent message.
+     *  - nonce: The nonce of the sent message.
+     *  - fee: The LayerZero fee incurred for the message.
+     */
+    function send(
+        SendParam calldata _sendParam,
+        MessagingFee calldata _fee,
+        address _refundAddress
+    ) public payable virtual override returns (MessagingReceipt memory msgReceipt, OFTReceipt memory oftReceipt) {
+        // @dev Ensure the native funds in msg.value are enough to cover the fees and amount to send (with dust removed).
+        // TODO should we remove dust from _sendParam.amountLD here?
+        if (_fee.nativeFee + _removeDust(_sendParam.amountLD) != msg.value) {
+            revert NotEnoughNative(msg.value);
```

Comment:
Wondering if instead of reverting with `NotEnoughNative` we should revert with a custom error that also indicates how much native _should_ have been provided (i.e `_fee.nativeFee + _removeDust(_sendParam.amountLD)`)

Reply:
it technically is the same purpose, but I could be convinced to do a separate error. The issue is that the original assertion for the fee->msg.value doesn't happen now, so the underlying behaviour is inconsistent

Reply:
Gotcha, yeah in that case I'll add a new custom error specifically for this use case. It will also help the user see what they should have passed for `msgValue` if they accidentally pass an incorrect value

Reply:
Actually, I think we should reuse `InsufficientMessageValue` error here as that already lets us indicate `provided` and `required` and fits this use case